pkg/engine/internal/scheduler/scheduler_test.go

```diff
@@ -0,0 +1,1249 @@
+package scheduler
+
+import (
+	"context"
+	"os"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/dskit/services"
+	"github.com/oklog/ulid/v2"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/atomic"
+
+	"github.com/grafana/loki/v3/pkg/engine/internal/executor"
+	"github.com/grafana/loki/v3/pkg/engine/internal/planner/physical"
+	"github.com/grafana/loki/v3/pkg/engine/internal/scheduler/wire"
+	"github.com/grafana/loki/v3/pkg/engine/internal/workflow"
+)
+
+func TestScheduler_AddStreams(t *testing.T) {
+	t.Run("Succeeds with new stream", func(t *testing.T) {
+		sched := newTestScheduler(t)
+
+		stream := workflow.Stream{ULID: ulid.Make()}
+		require.NoError(t, sched.AddStreams(t.Context(), nopStreamHandler, &stream))
+	})
+
+	t.Run("Fails with existing stream", func(t *testing.T) {
+		sched := newTestScheduler(t)
+
+		stream := workflow.Stream{ULID: ulid.Make()}
+		require.NoError(t, sched.AddStreams(t.Context(), nopStreamHandler, &stream))
+		require.Error(t, sched.AddStreams(t.Context(), nopStreamHandler, &stream), "Scheduler should reject existing stream")
+	})
+
+	t.Run("Fails with zero ULID", func(t *testing.T) {
+		sched := newTestScheduler(t)
+
+		stream := workflow.Stream{ULID: ulid.Zero}
+		require.Error(t, sched.AddStreams(t.Context(), nopStreamHandler, &stream), "Scheduler should reject zero-value ULID")
+	})
+}
+
+func TestScheduler_RemoveStreams(t *testing.T) {
+	t.Run("Fails with unrecognized stream", func(t *testing.T) {
+		sched := newTestScheduler(t)
+
+		stream := workflow.Stream{ULID: ulid.Make()}
+		require.Error(t, sched.RemoveStreams(t.Context(), &stream), "Scheduler should reject unrecognized stream")
+	})
+
+	t.Run("Succeeds with recognized stream", func(t *testing.T) {
+		sched := newTestScheduler(t)
+
+		stream := workflow.Stream{ULID: ulid.Make()}
+		require.NoError(t, sched.AddStreams(t.Context(), nopStreamHandler, &stream))
+		require.NoError(t, sched.RemoveStreams(t.Context(), &stream))
+	})
+
+	t.Run("Fails with removed stream", func(t *testing.T) {
+		sched := newTestScheduler(t)
+
+		stream := workflow.Stream{ULID: ulid.Make()}
+		require.NoError(t, sched.AddStreams(t.Context(), nopStreamHandler, &stream))
+		require.NoError(t, sched.RemoveStreams(t.Context(), &stream))
+		require.Error(t, sched.RemoveStreams(t.Context(), &stream), "Scheduler should reject already removed stream")
+	})
+
+	t.Run("Fails with stream that has active tasks", func(t *testing.T) {
+		sched := newTestScheduler(t)
+
+		stream := workflow.Stream{ULID: ulid.Make()}
+		require.NoError(t, sched.AddStreams(t.Context(), nopStreamHandler, &stream), "Scheduler should accept new stream")
+
+		var (
```

Comment:
we don't need a `var` block here, can we just assign receiver and sender with `:=`?